docs/GraphRepresentation.md

```diff
@@ -0,0 +1,321 @@
+in the resulting matrix.
+In terms of the graph representation,
+the weight on each edge (A,B) leaving A is first divided by
+the sum of the weights on all edges leaving A,
+and then the resulting edge weight is divided by
+the sum of the weights on all edges entering B.
+
+Intuitively, the first step converts coinstallation counts for A
+to the proportion of coinstallations of A allocated to each add-on B.
+In the case of general relevance scores, we can think of the row sum for A
+as a __relevance budget__,
+ie. the total amount of relevance it has available to assign to other add-ons.
+The first division has the effect of normalizing each add-on's
+relevance budget to 1.
+The next step then normalizes each coinstalled add-on's
+resulting aggregate relevance to 1.
+
+This treatment is implemented as [`RowNormSum`](../taar_lite/recommenders/treatments.py#L126).
+
+
+## Graph pruning
+
+Another class of treatments we may wish to apply involves modifying
+the vertices or edges of the relational graph themselves,
+possibly as a function of the edge weights.
+For example, we may wish to remove vertices or edges that fall below
+a minimum relevance threshold, in order to avoid recommending rare add-ons.
+Also, we may wish to ensure certain properties of the graph as a whole.
+It is possible for a user visiting AMO to "walk" the recommendation graph
+by clicking through a chain of subsequent recommendations.
+Although the recommendations themselves are memoryless,
+in that they only depend on the current add-on,
+we are interesting in optimizing the quality of experience
+for such users as well,
+such as not returning to a previous add-on in a short number of steps.
+
+
+### Minimum relevance pruning
+
+This treatment removes add-on vertices whose relevance score is lower
+than a given threshold,
+under the assumption that candidates with such low relevance
+would not make good recommendations, even if there no candidates
+with higher relevance.
+
+The relevance threshold is computed based on an auxiliary list
+of overall relevance scores for each add-on.
+It is chosen as 5% of the mean of all the overall scores.
+Then, if any add-on has a relevance score below this value,
+its vertex (and all incident edges) is removed from the relational graph.
+Equivalently, the corresponding row and column are removed from the matrix $C$.
+
+This treatment is implemented as [`MinInstallPrune`](../taar_lite/recommenders/treatments.py#L36).
+The auxiliary score list is supplied as the keyword arg `ranking_dict`,
+a dict mapping add-on GUIDs to overall relevance scores.
+
+
+### Recommendation selection
+
+The procedure for [selecting recommendations](./GuidGuidRecommender.md#selecting-recommendations)
+it itself just another graph pruning treatment.
+
+For each add-on A in the graph, the edges leaving A are ordered
+by decreasing relevance score (weight),
+with ties broken using a supplied list of auxiliary add-on scores.
```

Comment:
Ties could be broken in a bunch of ways. Is this a general paragraph or an implementation description?

Reply:
It's an implementation description, as are all the treatment descriptions in this doc. The general overview of the recommendation process is in the GuidGuidRecommender doc.